### ISSUE TITLE
meshConfig.accessLogFormat doesn't work

### ISSUE THREAD

**Issue by jfassad** [istio/istio#24318]
**Bug description**

setting meshConfig.accessLogFormat has no effect. had to fallback to global.proxy.accessLogFormat

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

Setting meshConfig.accessLogFormat should change the log format

**Steps to reproduce the bug**

This configuration doesn't change log format:
```
  meshConfig:
    accessLogEncoding: JSON
    accessLogFile: "/dev/stdout"
    accessLogFormat: |
      {
        "log_type": "access_log",
        "start_time": "%START_TIME%",
        "route_name": "%ROUTE_NAME%",
        "method": "%REQ(:METHOD)%",
        "path": "%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%",
        "protocol": "%PROTOCOL%",
        "response_code": "%RESPONSE_CODE%",
        "response_flags": "%RESPONSE_FLAGS%",
        "bytes_received": "%BYTES_RECEIVED%",
        "bytes_sent": "%BYTES_SENT%",
        "duration": "%DURATION%",
        "upstream_service_time": "%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%",
        "x_forwarded_for": "%REQ(X-FORWARDED-FOR)%",
        "user_agent": "%REQ(USER-AGENT)%",
        "request_id": "%REQ(X-REQUEST-ID)%",
        "authority": "%REQ(:AUTHORITY)%",
        "upstream_host": "%UPSTREAM_HOST%",
        "upstream_cluster": "%UPSTREAM_CLUSTER%",
        "upstream_local_address": "%UPSTREAM_LOCAL_ADDRESS%",
        "downstream_local_address": "%DOWNSTREAM_LOCAL_ADDRESS%",
        "downstream_remote_address": "%DOWNSTREAM_REMOTE_ADDRESS%",
        "requested_server_name": "%REQUESTED_SERVER_NAME%",
        "istio_policy_status": "%DYNAMIC_METADATA(istio.mixer:status)%",
        "upstream_transport_failure_reason": "%UPSTREAM_TRANSPORT_FAILURE_REASON%"
      }
```
This works:
```
  meshConfig:
    enablePrometheusMerge: false
    accessLogEncoding: JSON
    accessLogFile: "/dev/stdout"

  values:
    global:
      proxy:
        accessLogFormat: |
          {
            "log_type": "access_log",
            "start_time": "%START_TIME%",
            "route_name": "%ROUTE_NAME%",
            "method": "%REQ(:METHOD)%",
            "path": "%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%",
            "protocol": "%PROTOCOL%",
            "response_code": "%RESPONSE_CODE%",
            "response_flags": "%RESPONSE_FLAGS%",
            "bytes_received": "%BYTES_RECEIVED%",
            "bytes_sent": "%BYTES_SENT%",
            "duration": "%DURATION%",
            "upstream_service_time": "%RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%",
            "x_forwarded_for": "%REQ(X-FORWARDED-FOR)%",
            "user_agent": "%REQ(USER-AGENT)%",
            "request_id": "%REQ(X-REQUEST-ID)%",
            "authority": "%REQ(:AUTHORITY)%",
            "upstream_host": "%UPSTREAM_HOST%",
            "upstream_cluster": "%UPSTREAM_CLUSTER%",
            "upstream_local_address": "%UPSTREAM_LOCAL_ADDRESS%",
            "downstream_local_address": "%DOWNSTREAM_LOCAL_ADDRESS%",
            "downstream_remote_address": "%DOWNSTREAM_REMOTE_ADDRESS%",
            "requested_server_name": "%REQUESTED_SERVER_NAME%",
            "istio_policy_status": "%DYNAMIC_METADATA(istio.mixer:status)%",
            "upstream_transport_failure_reason": "%UPSTREAM_TRANSPORT_FAILURE_REASON%"
          }
```

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

istioctl version --remote
client version: 1.6.0
control plane version: 1.6.0
data plane version: 1.6.0 (9 proxies)

kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:56:40Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.9", GitCommit:"a17149e1a189050796ced469dbd78d380f2ed5ef", GitTreeState:"clean", BuildDate:"2020-04-16T11:36:15Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}

Not using helm

**How was Istio installed?**
I've tried this on three different clusters.
Two were installed with istioctl and the third with istio-operator controller

**Environment where bug was observed (cloud vendor, OS, etc)**
aws kops 1.16.2 - ubuntu 20.04

**Comment by howardjohn** [comment]
```
    accessLogFormat: |
      {
        "a": "b"
      }
```

fails

```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
spec:
  hub: gcr.io/istio-testing
  tag: latest

  # You may override parts of meshconfig by uncommenting the following lines.
  meshConfig:
    accessLogEncoding: JSON
    accessLogFile: "/dev/stdout"
    accessLogFormat: |
      {
        
      }
```

succeeds. 

Thanks for the report, will investigate

**Comment by jdomag** [comment]
@howardjohn 
Any update on this?
Same happens in istio 1.6.5 - the upgrade from 1.5 broke our logging capabilities

**Comment by samirspatel** [comment]
Don't mean to bug ya @howardjohn but we are having this exact issue with istio 1.6 and need a resolution. I cannot even get the json parameter to work properly so defining the format explicitly is the only recourse.

**Comment by winterTTr** [comment]
After soo many experiment and retry today, i finally figure out a really hacky workaroud, i believe this definitely a bug some where, but hopefully this work around can help someone temporarily: 

For istio 1.7, I found istio even remove the "values.global.proxy.accessLogFormat", so only below seems to work now:

```yaml
  meshConfig:
    accessLogFile: /dev/stdout
    accessLogEncoding: JSON
    accessLogFormat: |
      {
        "accessLogFormat": "{\"a\": \"access_log\",\"start_time\": \"%START_TIME%\"}"
      }
```

This will generate a log from your application sidecar and ingress gateway as:
```json
{"start_time":"2020-09-08T07:16:08.326Z","a":"access_log"}
```

**Comment by jdomag** [comment]
@winterTTr 
Not a fix and looks super ugly in the config, but works! :) thanks, that's much better than patching configmap manually

**Comment by winterTTr** [comment]
> @winterTTr
> Not a fix and looks super ugly in the config, but works! :) thanks, that's much better than patching configmap manually

Yes, this definitely should be fixed from istio side :-\

**Comment by Ghazgkull** [comment]
This defect has only gotten worse since it was reported. Anything we can do to reopen it?

**Comment by winterTTr** [comment]
Yes, this is very basic configuration, and i believe every one who really use the istio in production will need it. Why no one from istio team join and take care of this since 1.6.

**Comment by Ghazgkull** [comment]
@howardjohn Does the Istio project have any way to flag regressions like this? This is something that worked and was broken by a defect that snuck into a release.

**Comment by ostromart** [comment]
The problem comes from the fact that translate uses the tpath package, which treats YAML strings as objects for the purpose of supporting overlaying trees. yaml.Unmarshal unmarshals valid JSON strings as YAML which is not expected. The workaround is to check if the string is JSON and if so, treat it as a string. 
This still leaves a potential hole of paths that expect a YAML to be passed as string. There are no such paths currently but the correct (though quite heavy-handed) solution to this would be to pass an Options into WriteNode to configure its behavior.

**Comment by ostromart** [comment]
Fixed, also backported to 1.7. Test included to prevent any regressions.

**Comment by qianweichun** [comment]
@ostromart Could you please merge the fix to 1.6?

**Comment by howardjohn** [comment]
1.6 us EOL and won't get any more releases
